dkc/core/rest/file.py

```diff
@@ -109,3 +113,23 @@ def download(self, request, pk=None):
         if file.blob:  # FieldFiles are falsy when not populated with a file
             return HttpResponseRedirect(file.blob.url)
         return Response(status=204)
+
+    @swagger_auto_schema(
+        query_serializer=HashDownloadSerializer,
+        responses={
+            200: None,  # needed to override default
+            302: 'You will be redirected to download the file contents.',
+            404: 'No file exists with the given hash.',
+        },
+    )
+    @action(detail=False, filterset_fields=[], pagination_class=None)
+    def hash_download(self, request):
+        """Download a file based on the hash of its contents."""
+        serializer = HashDownloadSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        sha512 = serializer.validated_data['sha512'].lower()
+        qs = self.get_queryset().filter(sha512=sha512).only('blob').order_by()
```

Comment:
I added a little bit of extra tweaking to the queryset to slightly optimize the underlying query. Since hash download seems to be the bulk of the workload taking place on DKC, it seems worth this small amount of extra effort.